[TnT-85] 1차 컴포넌트 구현

SeonJeongk opened the pull request:
## 📝 작업 내용

- Closes #12 
<br>

- 1차로 확정된 컴포넌트들을 구현했습니다.
  >  Button,TextField, Dialog, TopBar

- **Button**은 `ButtonUtils`의  **getButtonConfig**를 통해 `Size`(XLarge, Large, Medium, ...) 와 `Type`(Primary, Gray, GrayOutline, RedOutline), enabled에 맞는 설정을 넘겨받습니다.
- **TopBar**는 `뒤로가기만`, `뒤로가기 + 제목`, `뒤로가기 + 제목 + component`의 3가지 옵션이 있습니다.
- **TextField**는 입력란과 right component, 에러문구를 보여주는 `TnTTextField`와  제목과  counter까지 보여주는 `TnTLabeledTextField`로 구현했습니다.

- 추가로 기존 Typo와 Color를 추가 및 수정했습니다.

## 📸 실행 화면
<img src="https://github.com/user-attachments/assets/e9ef35b4-367d-47e5-a486-11874756a587" width="230"/>
<img src="https://github.com/user-attachments/assets/da35a7b0-7bc3-4f15-b4ca-e569995fc913" width="230"/>
<img src="https://github.com/user-attachments/assets/43b4f307-f531-4e3e-825a-40638eda8115" width="230"/>
<img src="https://github.com/user-attachments/assets/54531eb1-9112-47fe-a47f-90a4dcaf7ec8" width="230"/>




## 🙆🏻 리뷰 요청 사항
없습니다


## 👀 레퍼런스
[버튼 최소 크기 수정](https://stackoverflow.com/questions/69156462/how-to-remove-or-reduce-padding-in-jetpack-compose-outlinedbutton)
[TextField 커스텀](https://reco-dy.tistory.com/16)

SeonJeongk committed:
[TNT-85] design: Typography 수정

SeonJeongk committed:
[TNT-85] design: 색상 추가 및 Theme 수정

SeonJeongk committed:
[TNT-85] design: TopBar 컴포넌트 구현

SeonJeongk committed:
[TNT-85] design: Red_50 색상 추가

SeonJeongk committed:
[TNT-85] fix: 미사용 코드 삭제 (darkTheme)

SeonJeongk committed:
[TNT-85] design: Button 컴포넌트 구현

- TextButton, IconButton, BottomButton
- ButtonUtils에서 버튼 설정 관리

SeonJeongk committed:
[TNT-85] design: TextField 컴포넌트 구현

SeonJeongk committed:
[TNT-85] design: PopupDialog 컴포넌트 구현

SeonJeongk committed:
[TNT-85] design: LabeledTextField optional 추가

hoyahozz commented on core/designsystem/src/main/java/co/kr/tnt/designsystem/util/ButtonUtils.kt:
```diff
@@ -0,0 +1,132 @@
+import androidx.compose.ui.unit.dp
+import co.kr.tnt.designsystem.theme.TnTTheme
+
+enum class ButtonSize {
+    XLarge,
+    Large,
+    Medium,
+    Small,
+    XSmall,
+}
```
요기 있는 클래스들 모두 패키징이 필요해보입니다!

`component > button > model` 로 패키징하는거 어떨까요?

SeonJeongk replied:
넵 수정해두겠습니다!

hoyahozz commented on core/designsystem/src/main/java/co/kr/tnt/designsystem/util/ButtonUtils.kt:
```diff
@@ -0,0 +1,132 @@
+)
+
+@Composable
+fun getButtonConfig(size: ButtonSize, variant: ButtonType, enabled: Boolean): ButtonConfig {
```
`ButtonConfig` 를 따로 분리하신 이유가 있으실까요?

제 생각에 `height`, `contentPadding` 과 같은 필드들은 `ButtonSize` 에서,

`color`, `stroke` 와 같은 필드들은 `ButtonType` 내의 프로퍼티 혹은 메소드로 분리할 수 있을 것처럼 보여요!

SeonJeongk replied:
모든 정보를 한 곳에서 처리하려고 하다보니 ButtonConfig로 분리했습니다
각 클래스에 관련된 속성들을 분리하여 관리하는게 더 좋아보이네요! 수정해보겠습니다👍

hoyahozz commented on core/designsystem/src/main/java/co/kr/tnt/designsystem/component/Button.kt:
```diff
@@ -0,0 +1,202 @@
+
+@Composable
+fun TnTTextButton(
+    size: ButtonSize = ButtonSize.Medium,
+    type: ButtonType = ButtonType.Primary,
+    text: String,
+    enabled: Boolean = true,
+    onClick: () -> Unit,
+    modifier: Modifier = Modifier,
```
컴포즈에서는 권장되는 파라미터 순서가 있어요!

일반적으로 필수 파라미터 (`text`)는 가장 최상위에, 옵셔널한 파라미터는 그 하위에 작성하도록 권장하고 있습니다!

그 중에서도, `modifier` 는 옵셔널한 파라미터 중 최상단에 위치하기를 권장하고 있어요!

> [Since the modifier is recommended for any component and is used often, placing it first ensures that it can be set without a named parameter and provides a consistent place for this parameter in any component.](https://android.googlesource.com/platform/frameworks/support/+/androidx-main/compose/docs/compose-component-api-guidelines.md#parameter)

SeonJeongk replied:
안 그래도 구현하면서 순서에 대해 고민했습니다.. ;-; 
첨부 링크 참고해서 수정해보겠습니다!

hoyahozz commented on core/designsystem/src/main/java/co/kr/tnt/designsystem/component/Button.kt:
```diff
@@ -0,0 +1,202 @@
+                text = text,
+                style = TnTTheme.typography.h4,
+            )
+        }
```
```suggestion
        Text(
            text = text,
            style = TnTTheme.typography.h4,
            modifier = Modifier.align(Alignment.Top)
        )
```

`Column` 안감싸고 `Text` 로만 하위 컴포저블 구현 가능할 것 같아요 !_!

SeonJeongk replied:
오매 수정했습니다 감사합니다!

hoyahozz commented on core/designsystem/src/main/java/co/kr/tnt/designsystem/component/TextField.kt:
```diff
@@ -0,0 +1,240 @@
+    isSingleLine: Boolean = false,
+    showWarning: Boolean = false,
+    warningMessage: String? = null,
+    rightComponent: @Composable () -> Unit = {},
```
우리 버튼 포지션에서 `Trailing`, `Leading` 이라는 표현을 썼으니까 요기서도 적용해봐요!

hoyahozz commented on core/designsystem/src/main/java/co/kr/tnt/designsystem/component/TextField.kt:
```diff
@@ -0,0 +1,240 @@
+                .padding(bottom = 7.dp, end = 4.dp)
+                .fillMaxWidth(),
+        ) {
+            Row {
+                Text(
+                    text = title,
+                    style = TnTTheme.typography.body1Bold,
+                    color = TnTTheme.colors.neutralColors.Neutral900,
+                )
+                if (!optional) {
+                    Text(
+                        text = "*",
+                        style = TnTTheme.typography.body1Bold,
+                        color = TnTTheme.colors.mainColors.Red500,
+                    )
+                }
+            }
+
+            Text(
+                text = "${value.length}/$maxLength",
+                style = TnTTheme.typography.label1Medium,
+                color = counterColor,
+            )
```
```suggestion
            Text(
                text = title,
                style = TnTTheme.typography.body1Bold,
                color = TnTTheme.colors.neutralColors.Neutral900,
            )
            if (!optional) {
                Text(
                    text = "*",
                    style = TnTTheme.typography.body1Bold,
                    color = TnTTheme.colors.mainColors.Red500,
                )
            }
            Spacer(Modifier.weight(1f))
            Text(
                text = "${value.length}/$maxLength",
                style = TnTTheme.typography.label1Medium,
                color = counterColor,
            )
```

끝으로 제한 글자 수 위치시키기 위함이라면, 위와 같이 `Spacer` 를 활용할 수 있을 것 같아요!

SeonJeongk replied:
옹 `Spacer`가 있었네요 수정했습니다!

hoyahozz commented on core/designsystem/src/main/java/co/kr/tnt/designsystem/component/TextField.kt:
```diff
@@ -0,0 +1,240 @@
+            horizontalArrangement = Arrangement.SpaceBetween,
+            verticalAlignment = Alignment.CenterVertically,
+            modifier = Modifier
+                .padding(bottom = 7.dp, end = 4.dp)
```
요거 8dp 로 변경됐어여 !

hoyahozz commented on core/designsystem/src/main/java/co/kr/tnt/designsystem/component/Button.kt:
```diff
@@ -0,0 +1,202 @@
+        contentPadding = config.contentPadding,
+        modifier = modifier
+            .height(config.height)
+            .defaultMinSize(minWidth = 1.dp, minHeight = 1.dp),
```
요친구는 왜 필요한건가영 ?.?

SeonJeongk replied:
`Small`과 `XSmall` 사이즈 버튼은 Button의 기본 `MinWidth`보다 작은 경우가 있기 때문에 이를 고려해 해당 부분을 추가했습니다! 

혹시 다른 방식으로 구현할 수 있을까요?

#### 1. `defaultMinSize`를 설정하지 않았을 때 Small, XSmall 사이즈 버튼
<img src="https://github.com/user-attachments/assets/b21eaec6-167d-471b-b537-b7c4989048d7" width="100"/>

#### 2. `defaultMinSize`를 설정했을 때 Small, XSmall 사이즈 버튼
<img src="https://github.com/user-attachments/assets/d11ed202-08e1-4d26-b3a0-87f9d06f2b6c" width="100"/>

<br>
<br>

> 참고 : https://stackoverflow.com/questions/69156462/how-to-remove-or-reduce-padding-in-jetpack-compose-outlinedbutton

hoyahozz replied:
오호 그럼 XSmallSize 의 width, height 로 설정해줄 수 있을 것 같아요!

hoyahozz commented on core/designsystem/src/main/java/co/kr/tnt/designsystem/component/Button.kt:
```diff
@@ -0,0 +1,202 @@
+        enabled = enabled,
+        colors = config.colors,
+        shape = RoundedCornerShape(config.cornerRadius),
+        border = BorderStroke(width = config.stroke, color = config.borderColor),
```
```suggestion
        border = BorderStroke(
            width = config.stroke,
            color = config.borderColor,
        ),
```

취향차이긴 한데 파라미터 두 개이상인 경우 개행해주는거 어떨까요 ?.?

SeonJeongk replied:
옹 코드가 길어지는 것 때문에 개행하지 않고 구현했습니다
개행하는거 좋아요~! 수정해두겠습니다

hoyahozz commented on core/designsystem/src/main/java/co/kr/tnt/designsystem/component/TextField.kt:
```diff
@@ -0,0 +1,240 @@
+
+@Preview(showBackground = true, heightDp = 100)
+@Composable
+fun TnTTextFieldPreview() {
```
모든 프리뷰 컴포저블 함수는 `private` 하게 선언해주세요!

hoyahozz commented on core/designsystem/src/main/java/co/kr/tnt/designsystem/component/TextField.kt:
```diff
@@ -0,0 +1,240 @@
+@Composable
+fun TnTTextField(
+    value: String,
+    placeholder: String,
```
`placeHolder` 도 `nullable` 한 타입으로 선언해주는건 어떨까요? 

필요하지 않은 경우도 있을 것 같아서요!

hoyahozz commented on core/designsystem/src/main/java/co/kr/tnt/designsystem/component/TextField.kt:
```diff
@@ -0,0 +1,240 @@
+                    .padding(end = 8.dp),
+            ) {
+                if (value.isEmpty()) {
+                    Text(
```
오호 이렇게 구현하는 것도 괜찮기는 한데, `BasicTextField` 내 `decorationBox` 활용하면, `PlaceHolder` 구현 가능해요!

참고 : https://reco-dy.tistory.com/16

SeonJeongk replied:
옹 넵 참고해서 수정했습니다!!

hoyahozz commented on core/designsystem/src/main/java/co/kr/tnt/designsystem/component/TopBar.kt:
```diff
@@ -0,0 +1,127 @@
+// back button
+@Preview(showBackground = true)
+@Composable
+fun TnTTopBarPreview1() {
```
1, 2, 3 숫자 대신 좀 더 의미있는 함수명을 작성해봅시당!

`TnTTopBarOnlyBackButtonPreview` 와 같은 이름은 어때요? ㅎㅎ

SeonJeongk replied:
`TnTTopBarOnlyBackButtonPreview`, `TnTTopBarBackButtonWithTitlePreview`, `TnTTopBarWithAllComponentsPreview`로 수정했습니다!

hoyahozz reviewed (changes requested):
디자인 시스템 구현하시느라 너무 고생하셨어요 :->

코멘트 남겨놓았으니 한 번 확인해주세요 ~!

SeonJeongk committed:
[TNT-85] move: button 패키지 생성, Button 이동

SeonJeongk committed:
[TNT-85] refactor: Button class 패키징 및 BuildConfig 삭제

- 파라미터 2개 이상인 경우 개행 추가
- BuildConfig의 역할을 각 클래스 내 메소드로 분리
- [size] height, contentPadding, cornerRadius, textStyle
- [type] stoke, borderColor, button colors

SeonJeongk committed:
[TNT-85] refactor: 권장 사항에 맞춰 파라미터 순서 수정

SeonJeongk committed:
[TNT-85] refactor: TnTBottomButton 레이아웃 수정

SeonJeongk committed:
[TNT-85] refactor: Preview private으로 수정

- TopBar Preview 함수명 변경

SeonJeongk committed:
[TNT-85] refactor: 변수명 trailingComponent로 수정

- textField 갭 8.dp로 수정

SeonJeongk committed:
[TNT-85] refactor: placeholder 위치 수정 및 nullable로 변경

SeonJeongk committed:
[TNT-85] refactor: TextField 디자인 변경 반영, Spacer 추가

- 입력란 padding 8.dp로 수정
- 아래 라인 색상 Neutral200으로 변경

SeonJeongk committed:
[TNT-85] feat: single button popup dialog 구현

SeonJeongk committed:
[TNT-85] fix: 트레일링 콤마 추가

SeonJeongk committed:
[TNT-85] feat: OutlinedTextField 추가

SeonJeongk commented:
리뷰 남겨주신 부분 모두 반영했습니다! 확인 한 번 부탁드립니다👍 @hoyahozz 
추가로 회원가입 화면에 쓰이는 아래 두 컴포넌트 `TnTOutlinedTextField`와 `TnTSingleButtonPopupDialog`도 추가로 구현했습니다 'o'

<img src="https://github.com/user-attachments/assets/d78ba3d3-5c2e-4a3b-9e3a-ea652657a2c2" width="150"/><img src="https://github.com/user-attachments/assets/bdef707d-214f-4f35-835e-ebc508978087" width="150"/>

SeonJeongk committed:
[TNT-85] fix: 트레일링 콤마 추가

hoyahozz commented on core/designsystem/src/main/java/co/kr/tnt/designsystem/component/button/Button.kt:
```diff
@@ -0,0 +1,195 @@
+        contentPadding = size.contentPadding,
+        modifier = modifier
+            .height(size.height)
+            .defaultMinSize(minWidth = 1.dp, minHeight = 1.dp),
```
```suggestion
            .defaultMinSize(minWidth = Dp.Hairline),
```

hoyahozz commented on core/designsystem/src/main/java/co/kr/tnt/designsystem/component/TextField.kt:
```diff
@@ -0,0 +1,342 @@
+    maxLength: Int = 15,
+    isSingleLine: Boolean = false,
+    showWarning: Boolean = false,
+    optional: Boolean = false,
```
```suggestion
    isRequired: Boolean = false,
```

hoyahozz commented on core/designsystem/src/main/java/co/kr/tnt/designsystem/component/TopBar.kt:
```diff
@@ -0,0 +1,124 @@
+    title: String = "",
+    onBackClick: () -> Unit = {},
+    modifier: Modifier = Modifier,
+    trailingComponent: @Composable () -> Unit = {},
```
```suggestion
    trailingComponent: @Composable RowScope.() -> Unit = {},
```

hoyahozz commented on core/designsystem/src/main/java/co/kr/tnt/designsystem/component/button/model/ButtonType.kt:
```diff
@@ -0,0 +1,66 @@
+import androidx.compose.ui.unit.dp
+import co.kr.tnt.designsystem.theme.TnTTheme
+
+enum class ButtonType(
```
`Recomposition` 카운트 지속적으로 모니터링하기

hoyahozz reviewed (approved):
디스코드로 얘기 나눴던 내용 적어둡니다 ~

hoyahozz commented on core/designsystem/src/main/java/co/kr/tnt/designsystem/component/TopBar.kt:
```diff
@@ -0,0 +1,124 @@
+    modifier: Modifier = Modifier,
+    trailingComponent: @Composable () -> Unit = {},
+) {
+    Box(
+        contentAlignment = Alignment.Center,
+        modifier = modifier
+            .fillMaxWidth()
+            .height(60.dp)
+            .padding(start = 16.dp, top = 20.dp, end = 16.dp, bottom = 12.dp),
+    ) {
+        IconButton(
+            onClick = onBackClick,
+            modifier = Modifier
+                .size(32.dp)
+                .align(Alignment.CenterStart),
+        ) {
+            Icon(
+                painter = painterResource(R.drawable.ic_arrow_left),
+                contentDescription = "Go back",
+            )
+        }
+
+        Text(
+            text = title,
+            style = TnTTheme.typography.h4,
+            color = TnTTheme.colors.neutralColors.Neutral900,
+            textAlign = TextAlign.Center,
+            maxLines = 1,
+            overflow = TextOverflow.Ellipsis,
+        )
+
+        Row(
+            modifier = Modifier.align(Alignment.CenterEnd),
+        ) {
+            trailingComponent()
+        }
+    }
```
```suggestion
    CenterAlignedTopAppBar(
        modifier = modifier,
        title = {
            Text(
                text = title,
                style = TnTTheme.typography.h4,
                color = TnTTheme.colors.neutralColors.Neutral900,
                textAlign = TextAlign.Center,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
        },
        actions = trailingComponent,
        navigationIcon = {
            IconButton(
                onClick = onBackClick,
                modifier = Modifier.size(32.dp)
            ) {
                Icon(
                    painter = painterResource(R.drawable.ic_arrow_left),
                    contentDescription = "Go back",
                )
            }
        }
    )
```

SeonJeongk committed:
[TNT-85] refactor: CenterAlignedTopAppBar로 수정

SeonJeongk committed:
[TNT-85] refactor: Button minWidth 설정 수정

SeonJeongk committed:
[TNT-85] refactor: TextField 수정

- trailingComponent에 BoxScope 적용
- optional isRequired로 변경
- if-else문 true 기준으로 작성
- OutlinedTextField 디자인 수정사항 반영
